### DISCUSSION TITLE
db_pre_request in private schema

### DISCUSSION THREAD

**Discussion by cam-mcevenue** [supabase/supabase#18510] [Questions] [answered]
I am trying to protect my db_pre_request function from public API using a private schema. Based on #3233 and [postgREST docs](https://postgrest.org/en/stable/tutorials/tut1.html?highlight=db-pre-request#bonus-topic-immediate-revocation).

I've tested this in both cloud and local development.

```
create schema "private";

grant usage on schema "private" to "authenticator";

alter default privileges for role "postgres" in schema "private" grant all on functions to "authenticator";

create or replace function private.test() returns void
  language plpgsql
  security definer
  as $$
begin
    if not (current_setting('request.jwt.claims', true)::jsonb ? 'email') then
        ---User not authenticated
        perform set_config('response.status', '418', true);
    else
        ---User is authenticated
        perform set_config('response.status', '200', true);
    end if;
    
end
$$;

alter role authenticator set pgrst.db_pre_request TO 'private.test';
notify pgrst,'reload config';
```
**Local Dev**

1. Completely breaks API Docs
2. Results in Postgres error `authenticator@postgres ERROR:  permission denied for schema test at character 8` when making any request via supabase-js

**Cloud**

1. Completely breaks API docs. 
2. Logs are down right now so I can't check them

--- 

Anyone successfully accomplished this before? Am I missing something in my permissions grants? 

Everything works as expected if I just move this function to the public schema ie.

```
create or replace function public.test() returns void
  language plpgsql
  security definer
  as $$
begin
    if not (current_setting('request.jwt.claims', true)::jsonb ? 'email') then
        ---User not authenticated
        perform set_config('response.status', '418', true);
    else
        ---User is authenticated
        perform set_config('response.status', '200', true);
    end if;
    
end
$$;

alter role authenticator set pgrst.db_pre_request TO 'test';
notify pgrst,'reload config';
```

But now the function can be accessed via the public API which I don't want.

**Comment by GaryAustin1** [comment, marked as answer]
You could test by granting anon or authenticated to the schema private.  Normally authenticator is running as those roles when it makes calls.  Of course if that works it defeats your goal.

**Reply by cam-mcevenue** [reply]
I figured out how to get this to work just before you commented :). Though I'm not exactly clear why. I think it has something to do with how `authenticator` switches into `anon` and `authenticated` roles.

In order to have the db_request_function in a private schema, you must grant the following:

```
create schema "private";

grant usage on schema private to authenticator, anon, authenticated;
alter default privileges for role "postgres" in schema "private" grant all on functions to authenticator, anon, authenticated;

create or replace function private.test() returns void
  language plpgsql
  security definer
  as $$
begin
    if not (current_setting('request.jwt.claims', true)::jsonb ? 'email') then
        ---User not authenticated
        perform set_config('response.status', '418', true);
    else
        ---User is authenticated
        perform set_config('response.status', '200', true);
    end if;
    
end
$$;

alter role authenticator set pgrst.db_pre_request TO 'private.test';
notify pgrst,'reload config';

```

@GaryAustin1 I don't think it defeats my goal, but I could be wrong lol correct me if I am. This hides it from the public API which prevents any user from just spamming that endpoint. The function doesn't actually return anything, but sets a property on the request `request.claims`. Doing this to get around the token refresh issues related to https://github.com/supabase-community/supabase-custom-claims

This adds overhead of fetching the permissions on every request, but I'd rather that than potentially outdated permissions and storing anything permissions related on the JWT.

I'll rethink this if I run into performance issues during testing, but I have to do something similar with RLS, so I'd rather store it in memory on the request, and then just access it where I need to in RLS policies or checks in functions

```
create or replace function postgrest.db_pre_request()
    returns void
as $$
    with jwt_claims as (
        select (coalesce(current_setting('request.jwt.claims', true), '{}'))::jsonb as data
    ),
    headers as (
        select jsonb(coalesce(current_setting('request.headers', true), '{}'))::jsonb as data
    ),
    user_permissions as (
        -- gets user permissions
    ),
    active_organization as (
       -- gets org the user is currently logged into
    ),
    custom_claims as (
        select
            jwt_claims.data ||
            headers.data ||
            user_permissions.data ||
            active_organization.data AS aggregated_data
            from 
                jwt_claims, 
                headers, 
                user_permissions, 
                active_organization
    )
    select set_config('request.claims'::text, (select aggregated_data::text from custom_claims), false );
$$ language sql security definer;
```

**Reply by GaryAustin1** [reply]
So this schema is NOT set in the API schemas to be accessible and still works?  If so then it can't be accessed directly.

**Reply by cam-mcevenue** [reply]
Correct,

This schema is NOT in the API schemas. 

I've been trying to get this to work for a few days, and one BIG gotcha is how you define the function, and what you do inside it. A few rules I've found through exploring the docs and testing to avoid issues.

1. Don't modify the database (insert, update, delete). Only manipulation should be to PostgRESTs [request and response settings](https://postgrest.org/en/stable/references/transactions.html#transaction-scoped-settings)
2. Don't return anything (ie. return void)
3. Dont execute select statements on any function or table that is not accessible to any of the above mentioned roles (or use security definer like I did)

If you don't follow those rules, PostgREST won't be happy. It will actually break the API docs functionality of PostgREST and there is very little to go off of in the logs of where the issue is. If your in local development will also cause the database to fail health checks when starting up `service not healthy: [supabase_rest_<app_name>]`. So it won't start up
